Port MASTG-TEST-0015: Testing the Purposes of Keys (android) (by @guardsquare) (#3525)
OWASP/owasp-mstg: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0xx1.md

---
platform: android
title: References to Asymmetric Key Pairs Used For Multiple Purposes
id: MASTG-TEST-0xx1
type: [static]
weakness: MASWE-0012

> profiles: [L1, L2, P]

Reviewer: ```suggestion
profiles: [L2]
```

   - **Signing / verification:** [`KeyProperties.PURPOSE_SIGN`](https://developer.android.com/reference/android/security/keystore/KeyProperties#PURPOSE_SIGN) | [`KeyProperties.PURPOSE_VERIFY`](https://developer.android.com/reference/android/security/keystore/KeyProperties#PURPOSE_VERIFY)
   - **Encryption / decryption:** [`KeyProperties.PURPOSE_ENCRYPT`](https://developer.android.com/reference/android/security/keystore/KeyProperties#PURPOSE_ENCRYPT) | [`KeyProperties.PURPOSE_DECRYPT`](https://developer.android.com/reference/android/security/keystore/KeyProperties#PURPOSE_DECRYPT)
   - **Wrapping / unwrapping keys:** [`KeyProperties.PURPOSE_WRAP_KEY`](https://developer.android.com/reference/android/security/keystore/KeyProperties#PURPOSE_WRAP_KEY)

In other words, ensure that the app doesn't create a key that combines purposes across roles (for example, signing **and** decrypting) which is potentially insecure and must be avoided.

## Observation

The output should contain a list of locations where asymmetric keys are created, along with backtraces showing each KeyPair instantiation

## Evaluation

> The test case fails if you find any keys with mixed roles as described in **Steps** section.

Reviewer: ```suggestion
The test case fails if you find any keys used for multiple purposes.
```
